ls1intum/Artemis
# [Tests] Cypress: Make multiple submissions in Programming Exercise test and provide basic test infrastructure

### Motivation and Context


The current programming exercise End-to-End test does not do any changes in the online code editor and as such does not test if the user can actually interact with the code editor. Furthermore the current programming exercise test is a monolith and tests way too much from the UI (creates test course, adds student, creates programming exercise, ...).

### Description

This PR adds functionality to the cypress programming exercise test so that the E2E test does multiple submissions (causing an build error, passing some tests and passing all tests). This is done entirely from the UI. Furthermore this PR splits the test into multiple tests, so that each test only verifies one functionality (course creation/deletion, manually adding students to a course and programming exercise creation/deletion/participation). I also added a general test infrastructure to encompass cypress requests and common ui selectors and code (pageobjects).

### Steps for Testing

You have the choice of executing the tests locally against a local/remote setup or you can simply execute the bamboo build plan, which I linked below.
**If you want to test against your local environment** you need to make sure that you pass a username template and password template for the "username" and "password" fields in the cypress.json. Currently the cypress tests don't support automatic creation of the test users. An example for a username template would be "fixture_tester_USERID" (USERID will be replaced by the tests).
If you need help with the configuration files just write me and I can help/send you a config file.

1. Configure cypress:
     - cypress.env (set adminUsername/password and username/password)
     - cypress.json (set the baseUrl to your test system of your choice)
2. Use `yarn cypress:run` to run the cypress tests in headless mode

Alternative:
1. Verify that the tests are running in the [test pipeline](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT)
